File: modules/01d-transformer/py/01_attention.py
"""
Task 1 🔴 — Multi-head self-attention with causal masking, FROM SCRATCH.

What you'll learn:
  - Scaled dot-product attention: the core operation of every transformer
  - Why we divide by sqrt(d_k) (keeps softmax gradients healthy)
  - Causal masking: how a decoder is stopped from "seeing the future"
  - Multi-head attention: split d_model into h independent subspaces, attend
    in each, concatenate, then project back

The math (README derives each step in plain English):

  Scaled dot-product attention, for queries Q, keys K, values V:

    scores = Q @ K.T / sqrt(d_k)          shape (n_q, n_k)
    scores = scores + mask                 mask is 0 or -inf per entry
    weights = softmax(scores)              row-wise, each row sums to 1
    output  = weights @ V                  shape (n_q, d_v)

  Causal mask (additive form): an (n, n) matrix that is 0 on and below the
  diagonal and -inf above it. Adding it before softmax drives the weight on
  every "future" position j > i to exactly 0.

  Multi-head attention with h heads and model dim d_model (d_k = d_model / h):
    - project X -> Q, K, V  (weights provided, seeded)
    - reshape each into h heads of width d_k
    - run scaled dot-product attention independently in each head
    - concatenate the h head outputs back to width d_model
    - apply an output projection W_o

No ML library. Only numpy.

How to run:
  uv run python modules/01d-transformer/py/01_attention.py

The harness builds seeded projection weights, runs the checks, and prints the
acceptance evidence. The three core functions are left as TODOs for you.
"""

from __future__ import annotations

import numpy as np

NEG_INF = -1e9  # a large negative number; "-inf" for masking purposes


# ---------------------------------------------------------------------------
# Core functions — implement these
# ---------------------------------------------------------------------------


def scaled_dot_product_attention(
    Q: np.ndarray, K: np.ndarray, V: np.ndarray, mask: np.ndarray | None = None
) -> tuple[np.ndarray, np.ndarray]:
    """
    Scaled dot-product attention.

    Args:
      Q : queries, shape (n_q, d_k)
      K : keys,    shape (n_k, d_k)
      V : values,  shape (n_k, d_v)
      mask : optional additive mask of shape (n_q, n_k). Entries are 0 (keep) or
             a large negative number (block). Added to the scores BEFORE softmax.

    Returns:
      (output, weights)
        weights : (n_q, n_k), each row a probability distribution summing to 1
        output  : (n_q, d_v), the attention-weighted mix of value vectors

    The softmax must be numerically stable: subtract the per-row max before exp().

    TODO: implement. Steps:
      - Form the raw scores as Q @ K.T, then scale by dividing by sqrt(d_k)
        (d_k = K.shape[-1]). Result is (n_q, n_k).
      - If a mask was passed, add it to the scores (it blocks entries before softmax).
      - Apply a row-wise stable softmax: subtract the per-row max
        (axis=-1, keepdims=True) before np.exp, then normalise each row by its sum
        (again axis=-1, keepdims=True) so every row sums to 1.
      - Multiply the weights by V to get the (n_q, d_v) output.
      - Return the (output, weights) pair.
    """
    # TODO: implement scaled dot-product attention
    raise NotImplementedError("TODO: implement scaled_dot_product_attention()")


def causal_mask(n: int) -> np.ndarray:
    """
    Build an (n, n) additive causal mask.

    Entry [i, j] is:
      0        if j <= i   (position i is allowed to attend to position j)
      -inf     if j >  i   (position i must NOT attend to a future position)

    Adding this to the pre-softmax scores makes every "future" weight exp(-inf)=0.

    Use NEG_INF (defined above) as the stand-in for -inf.

    TODO: implement. Start from a zeros (n, n) array and set the strictly-upper-
    triangular entries (the future, j > i) to NEG_INF, leaving the diagonal and
    below at 0. Hint: np.triu(..., k=1) selects entries strictly above the diagonal.
    """
    # TODO: implement causal mask
    raise NotImplementedError("TODO: implement causal_mask()")


def multi_head_attention(
    X: np.ndarray,
    Wq: np.ndarray,
    Wk: np.ndarray,
    Wv: np.ndarray,
    Wo: np.ndarray,
    num_heads: int,
    mask: np.ndarray | None = None,
) -> np.ndarray:
    """
    Multi-head self-attention.

    Args:
      X  : input, shape (n, d_model)
      Wq, Wk, Wv : projection weights, each shape (d_model, d_model)
      Wo : output projection, shape (d_model, d_model)
      num_heads : h; must divide d_model evenly. d_k = d_model / h.
      mask : optional additive mask of shape (n, n), applied inside every head.

    Returns:
      output, shape (n, d_model)

    TODO: implement. Steps:
      - Compute d_k = d_model // num_heads.
      - Project the input once into Q, K, V (each X @ W..., shape (n, d_model)).
      - For each head, take that head's contiguous block of d_k columns from Q, K, V
        (head h owns columns h*d_k .. (h+1)*d_k) and run
        scaled_dot_product_attention on those slices, passing the mask through.
        Keep just the output of each head.
      - Concatenate the per-head outputs along the feature axis (axis=1) back to
        width d_model (np.concatenate).
      - Apply the output projection Wo to the concatenation and return it.
    """
    # TODO: implement multi-head attention
    raise NotImplementedError("TODO: implement multi_head_attention()")


# ---------------------------------------------------------------------------
# Harness — complete, do not edit
# ---------------------------------------------------------------------------


def seeded_weights(d_model: int, seed: int) -> np.ndarray:
    """Deterministic projection matrix, Xavier-ish scale."""
    rng = np.random.default_rng(seed)
    scale = 1.0 / np.sqrt(d_model)
    return rng.normal(0.0, scale, (d_model, d_model)).astype(np.float64)


def main() -> None:
    rng = np.random.default_rng(0)
    n = 5  # sequence length
    d_model = 8  # model dimension
    h = 4  # number of heads (d_k = 2)

    X = rng.standard_normal((n, d_model))
    Wq = seeded_weights(d_model, 1)
    Wk = seeded_weights(d_model, 2)
    Wv = seeded_weights(d_model, 3)
    Wo = seeded_weights(d_model, 4)
    identity = np.eye(d_model)

    print("=" * 66)
    print("Task 1 — Multi-head self-attention with causal masking")
    print("=" * 66)
    print(f"  n={n}  d_model={d_model}  num_heads={h}  d_k={d_model // h}\n")

    # ── Check 1: attention weight rows sum to 1 ───────────────────────────────
    Q = rng.standard_normal((n, d_model))
    K = rng.standard_normal((n, d_model))
    V = rng.standard_normal((n, d_model))
    _, weights = scaled_dot_product_attention(Q, K, V)
    row_sums = weights.sum(axis=1)
    print("[1] Attention weight row sums:", np.round(row_sums, 6))
    rows_ok = np.allclose(row_sums, 1.0, atol=1e-6)
    print(f"    rows sum to 1 (+/-1e-6): {rows_ok}\n")

    # ── Check 2: causal mask zeroes out the future ────────────────────────────
    mask = causal_mask(n)
    _, cweights = scaled_dot_product_attention(Q, K, V, mask)
    # For every query position i, weight on any j > i must be ~0.
    upper = np.triu(cweights, k=1)
    causal_ok = np.allclose(upper, 0.0, atol=1e-9)
    print("[2] Causal attention weights (upper triangle should be all 0):")
    print(np.round(cweights, 3))
    print(f"    no attention to future positions (j>i == 0): {causal_ok}\n")

    # ── Check 3: h=1 multi-head equals single-head SDPA ───────────────────────
    # With identity projections and one head, MHA reduces to plain SDPA(X,X,X).
    mha_1 = multi_head_attention(X, identity, identity, identity, identity, num_heads=1)
    sdpa_out, _ = scaled_dot_product_attention(X, X, X)
    single_ok = np.allclose(mha_1, sdpa_out, atol=1e-9)
    print(f"[3] MHA(h=1, identity weights) == single-head SDPA: {single_ok}\n")

    # ── Check 4: shapes ───────────────────────────────────────────────────────
    out = multi_head_attention(X, Wq, Wk, Wv, Wo, num_heads=h, mask=causal_mask(n))
    shape_ok = out.shape == (n, d_model)
    print(f"[4] MHA output shape: {out.shape}  (expected ({n}, {d_model}))  ok={shape_ok}")
    print(f"    all outputs finite: {np.all(np.isfinite(out))}\n")

    assert rows_ok, "attention weight rows must sum to 1"
    assert causal_ok, "causal mask must zero out attention to future positions"
    assert single_ok, "MHA with h=1 and identity weights must equal single-head SDPA"
    assert shape_ok, "MHA output shape must be (n, d_model)"
    print("All checks passed. ✅")


if __name__ == "__main__":
    main()

File: modules/01d-transformer/ts/01-attention.ts
/**
 * Task 1 🔴 — Multi-head self-attention with causal masking (plain arrays, no ML lib).
 *
 * What you'll learn:
 *   - Scaled dot-product attention: the core operation of every transformer
 *   - Why we divide by sqrt(dK) (keeps softmax gradients healthy)
 *   - Causal masking: how a decoder is stopped from "seeing the future"
 *   - Multi-head attention: split dModel into h independent subspaces, attend
 *     in each, concatenate, then project back
 *
 * The math (README derives each step in plain English):
 *
 *   Scaled dot-product attention, for queries Q, keys K, values V:
 *     scores  = Q @ Kᵀ / sqrt(dK)          shape (nQ, nK)
 *     scores  = scores + mask               mask is 0 or -inf per entry
 *     weights = softmax(scores)             row-wise, each row sums to 1
 *     output  = weights @ V                 shape (nQ, dV)
 *
 *   Causal mask (additive form): an (n, n) matrix that is 0 on and below the
 *   diagonal and -inf above it. Adding it before softmax drives the weight on
 *   every "future" position j > i to exactly 0.
 *
 *   Multi-head attention with h heads and model dim dModel (dK = dModel / h):
 *     project X -> Q, K, V ; split into h heads ; attend per head ;
 *     concatenate the head outputs ; apply output projection Wo.
 *
 * No external math library. Plain number[][] arrays.
 *
 * How to run:
 *   pnpm tsx modules/01d-transformer/ts/01-attention.ts
 */

type Matrix = number[][];

const NEG_INF = -1e9; // a large negative number; "-inf" for masking purposes

// ---------------------------------------------------------------------------
// Matrix helpers (provided for you)
// ---------------------------------------------------------------------------

/** Matrix multiply: A (m×k) @ B (k×n) -> (m×n). */
function matMul(A: Matrix, B: Matrix): Matrix {
  const m = A.length,
    k = B.length,
    n = B[0].length;
  const out: Matrix = Array.from({ length: m }, () => new Array(n).fill(0) as number[]);
  for (let i = 0; i < m; i++)
    for (let p = 0; p < k; p++) {
      const aip = A[i][p];
      for (let j = 0; j < n; j++) out[i][j] += aip * B[p][j];
    }
  return out;
}

/** Transpose a matrix (n×m) -> (m×n). */
function transpose(A: Matrix): Matrix {
  return A[0].map((_, j) => A.map((row) => row[j]));
}

/** Slice columns [start, end) of every row. */
function sliceCols(A: Matrix, start: number, end: number): Matrix {
  return A.map((row) => row.slice(start, end));
}

/** Concatenate matrices side by side (same row count). */
function concatCols(mats: Matrix[]): Matrix {
  const rows = mats[0].length;
  const out: Matrix = [];
  for (let i = 0; i < rows; i++) {
    out.push(mats.flatMap((m) => m[i]));
  }
  return out;
}

// ---------------------------------------------------------------------------
// Core functions — implement these
// ---------------------------------------------------------------------------

/**
 * Scaled dot-product attention.
 *
 * @param Q queries, shape (nQ × dK)
 * @param K keys,    shape (nK × dK)
 * @param V values,  shape (nK × dV)
 * @param mask optional additive mask (nQ × nK): entries 0 (keep) or NEG_INF (block),
 *             added to the scores BEFORE softmax.
 * @returns { weights (nQ×nK, each row sums to 1), output (nQ×dV) }
 *
 * The softmax must be numerically stable: subtract the per-row max before exp().
 *
 * TODO: implement. Steps:
 *   - Form scores = matMul(Q, transpose(K)) and scale every entry by dividing by
 *     Math.sqrt(dK) (dK = K[0].length).
 *   - If a mask was passed, add its entries onto the scores (it blocks positions
 *     before the softmax).
 *   - Turn each score row into a probability distribution with a numerically-stable
 *     softmax: subtract that row's max before Math.exp, then divide by the row's sum
 *     so the row totals 1.
 *   - output = matMul(weights, V).
 *   - Return the { weights, output } object.
 */
function scaledDotProductAttention(
  Q: Matrix,
  K: Matrix,
  V: Matrix,
  mask: Matrix | null = null,
): { weights: Matrix; output: Matrix } {
  // TODO: implement scaled dot-product attention
  throw new Error("TODO: implement scaledDotProductAttention()");
}

/**
 * Build an (n × n) additive causal mask.
 *
 * Entry [i][j] is:
 *   0        if j <= i   (position i may attend to position j)
 *   NEG_INF  if j >  i   (position i must NOT attend to a future position)
 *
 * TODO: implement.
 *   Build an n×n array; for each i, j set NEG_INF when j > i, else 0.
 */
function causalMask(n: number): Matrix {
  // TODO: implement causal mask
  throw new Error("TODO: implement causalMask()");
}

/**
 * Multi-head self-attention.
 *
 * @param X  input, shape (n × dModel)
 * @param Wq @param Wk @param Wv projection weights, each (dModel × dModel)
 * @param Wo output projection, (dModel × dModel)
 * @param numHeads h; must divide dModel evenly. dK = dModel / h.
 * @param mask optional additive mask (n × n), applied inside every head.
 * @returns output, shape (n × dModel)
 *
 * TODO: implement. Steps:
 *   - Compute dK = dModel / numHeads.
 *   - Project X once into Q, K, V (each matMul(X, W...), n×dModel).
 *   - For each head, carve out that head's contiguous block of dK columns from Q, K, V
 *     with sliceCols (head hIdx owns columns hIdx*dK .. hIdx*dK+dK) and run
 *     scaledDotProductAttention on those slices, passing the mask through. Keep only
 *     each head's `output`.
 *   - Stitch the per-head outputs side by side with concatCols to get back to n×dModel.
 *   - Apply the output projection Wo (matMul) to the concatenation and return it.
 */
function multiHeadAttention(
  X: Matrix,
  Wq: Matrix,
  Wk: Matrix,
  Wv: Matrix,
  Wo: Matrix,
  numHeads: number,
  mask: Matrix | null = null,
): Matrix {
  // TODO: implement multi-head attention
  throw new Error("TODO: implement multiHeadAttention()");
}

// ---------------------------------------------------------------------------
// Harness — complete, do not edit
// ---------------------------------------------------------------------------

/** Deterministic Gaussian matrix (mulberry32 + Box-Muller), Xavier-ish scale. */
function seededWeights(rows: number, cols: number, seed: number): Matrix {
  let s = seed >>> 0;
  const scale = 1.0 / Math.sqrt(cols);
  const unif = () => {
    s = (s + 0x6d2b79f5) >>> 0;
    let t = s;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
  const gauss = () => {
    const u1 = unif() + 1e-12;
    const u2 = unif();
    return Math.sqrt(-2 * Math.log(u1)) * Math.cos(2 * Math.PI * u2) * scale;
  };
  return Array.from({ length: rows }, () => Array.from({ length: cols }, gauss));
}

function identity(n: number): Matrix {
  return Array.from({ length: n }, (_, i) =>
    Array.from({ length: n }, (_, j) => (i === j ? 1 : 0)),
  );
}

function allClose(A: Matrix, B: Matrix, atol = 1e-9): boolean {
  if (A.length !== B.length || A[0].length !== B[0].length) return false;
  for (let i = 0; i < A.length; i++)
    for (let j = 0; j < A[0].length; j++)
      if (Math.abs(A[i][j] - B[i][j]) > atol) return false;
  return true;
}

function round(x: number, d = 3): number {
  const f = 10 ** d;
  return Math.round(x * f) / f;
}

function main(): void {
  const n = 5,
    dModel = 8,
    h = 4;
  const X = seededWeights(n, dModel, 100);
  const Wq = seededWeights(dModel, dModel, 1);
  const Wk = seededWeights(dModel, dModel, 2);
  const Wv = seededWeights(dModel, dModel, 3);
  const Wo = seededWeights(dModel, dModel, 4);
  const I = identity(dModel);

  console.log("=".repeat(66));
  console.log("Task 1 — Multi-head self-attention with causal masking");
  console.log("=".repeat(66));
  console.log(`  n=${n}  dModel=${dModel}  numHeads=${h}  dK=${dModel / h}\n`);

  // ── Check 1: attention weight rows sum to 1 ─────────────────────────────────
  const Q = seededWeights(n, dModel, 10);
  const K = seededWeights(n, dModel, 11);
  const V = seededWeights(n, dModel, 12);
  const { weights } = scaledDotProductAttention(Q, K, V);
  const rowSums = weights.map((r) => r.reduce((a, b) => a + b, 0));
  console.log(
    "[1] Attention weight row sums:",
    rowSums.map((s) => round(s, 6)),
  );
  const rowsOk = rowSums.every((s) => Math.abs(s - 1) < 1e-6);
  console.log(`    rows sum to 1 (+/-1e-6): ${rowsOk}\n`);

  // ── Check 2: causal mask zeroes out the future ──────────────────────────────
  const mask = causalMask(n);
  const { weights: cweights } = scaledDotProductAttention(Q, K, V, mask);
  let causalOk = true;
  for (let i = 0; i < n; i++)
    for (let j = i + 1; j < n; j++)
      if (Math.abs(cweights[i][j]) > 1e-9) causalOk = false;
  console.log("[2] Causal attention weights (upper triangle should be all 0):");
  for (const row of cweights)
    console.log("   ", row.map((w) => round(w, 3)).join("  "));
  console.log(`    no attention to future positions (j>i == 0): ${causalOk}\n`);

  // ── Check 3: h=1 multi-head equals single-head SDPA ─────────────────────────
  const mha1 = multiHeadAttention(X, I, I, I, I, 1);
  const { output: sdpaOut } = scaledDotProductAttention(X, X, X);
  const singleOk = allClose(mha1, sdpaOut, 1e-9);
  console.log(`[3] MHA(h=1, identity weights) == single-head SDPA: ${singleOk}\n`);

  // ── Check 4: shapes ─────────────────────────────────────────────────────────
  const out = multiHeadAttention(X, Wq, Wk, Wv, Wo, h, causalMask(n));
  const shapeOk = out.length === n && out[0].length === dModel;
  const finite = out.every((r) => r.every((v) => Number.isFinite(v)));
  console.log(
    `[4] MHA output shape: ${out.length} x ${out[0].length}  (expected ${n} x ${dModel})  ok=${shapeOk}`,
  );
  console.log(`    all outputs finite: ${finite}\n`);

  console.assert(rowsOk, "attention weight rows must sum to 1");
  console.assert(causalOk, "causal mask must zero out attention to future positions");
  console.assert(
    singleOk,
    "MHA with h=1 and identity weights must equal single-head SDPA",
  );
  console.assert(shapeOk, "MHA output shape must be (n, dModel)");
  if (rowsOk && causalOk && singleOk && shapeOk) console.log("All checks passed. ✅");
}

if (import.meta.url === `file://${process.argv[1]}`) {
  main();
}
